sky/provision/paperspace/utils.py

```diff
@@ -0,0 +1,233 @@
+                                             data=search_kwargs)
+        items.extend(response['items'])
+        while response['hasMore']:
+            response = _try_request_with_backoff(
+                'get',
+                f'{API_ENDPOINT}/{endpoint}',
+                headers=self.headers,
+                data={
+                    'after': f'{response["nextPage"]}',
+                    **search_kwargs
+                })
+            items.extend(response['items'])
+        return items
+
+    def list_startup_scripts(
+            self, name: Optional[str] = None) -> List[Dict[str, Any]]:
+        return self.list_endpoint(
+            endpoint='startup-scripts',
+            name=name,
+        )
+
+    def get_sky_key_script(self) -> str:
+        return self.list_startup_scripts(ADD_KEY_SCRIPT)[0]['id']
+
+    def set_sky_key_script(self, public_key: str) -> None:
+        script = (
+            'if ! command -v docker &> /dev/null; then \n'
+            'apt-get update \n'
+            'apt-get install -y ca-certificates curl \n'
+            'install -m 0755 -d /etc/apt/keyrings \n'
+            'curl -fsSL https://download.docker.com/linux/ubuntu/gpg -o /etc/apt/keyrings/docker.asc \n'  # pylint: disable=line-too-long
+            'chmod a+r /etc/apt/keyrings/docker.asc \n'
+            'echo "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.asc] https://download.docker.com/linux/ubuntu \\\n'  # pylint: disable=line-too-long
+            '$(. /etc/os-release && echo "$VERSION_CODENAME") stable" | \\\n'
+            'tee /etc/apt/sources.list.d/docker.list > /dev/null \n'
+            'apt-get update \n'
+            'apt-get install -y docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin \n'  # pylint: disable=line-too-long
+            'fi \n'
+            'docker run hello-world \n'
+            'usermod -aG docker paperspace \n'
+            f'echo "{public_key}" >> /home/paperspace/.ssh/authorized_keys \n')
+        try:
+            script_id = self.get_sky_key_script()
+            _try_request_with_backoff(
+                'put',
+                f'{API_ENDPOINT}/startup-scripts/{script_id}',
+                headers=self.headers,
+                data=json.dumps({
+                    'name': ADD_KEY_SCRIPT,
+                    'script': script,
+                    'isRunOnce': True,
+                    'isEnabled': True
+                }))
+        except IndexError:
+            _try_request_with_backoff('post',
+                                      f'{API_ENDPOINT}/startup-scripts',
+                                      headers=self.headers,
+                                      data=json.dumps({
+                                          'name': ADD_KEY_SCRIPT,
+                                          'script': script,
+                                          'isRunOnce': True,
+                                      }))
```

Comment:
This is great to have a setup script setup for the instance creation! Could we move the call of this function to `config.bootstrap_instances` instead, where we can get the public key from the `config.node_config` to make it more clean for the ssh key setup function in `backend_utils`?

Reply:
Moved `set_sky_key_script` call to `bootstrap_instances` and made it so `setup_paperspace_authentication` only sets the pubkey path in `config`